### PR TITLE
Fix finish Allocation

### LINKED ISSUE #2624: Round Responded in Challenge Table is Always 0

### PR DESCRIPTION
**Pull request by Jayashsatolia403**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-july-1  |
| gosdk  | sprint-july-1  |
| zboxcli  | sprint-july-1  |
| zwalletcli  | sprint-july-1  |
| 0chain  | fix/finishAllocation  |
| blobber  | sprint-july-1  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | sprint-july-1  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

File: .github/workflows/benchmark-test.yml [hunk 1]
```diff
@@ -1,10 +1,15 @@
 name: Benchmark Tests
 
+concurrency:
```
Comment (dabasov) [inline .github/workflows/benchmark-test.yml]: do we need these changes?

File: .github/workflows/build-&-publish-docker-image.yml [hunk 2]
```diff
@@ -6,7 +6,7 @@ concurrency:
 
 on:
   push:
-    branches: [ master,staging ]
+    branches: [ master, staging, sprint* ]
```
Comment (dabasov) [inline .github/workflows/build-&-publish-docker-image.yml]: do we need to change this file

File: .github/workflows/system_tests.yml [hunk 3]
```diff
@@ -3,61 +3,30 @@ name: "0Chain System Tests"
...
-         default: 'staging'
-         required: true
-       zs3server_branch:
-         description: '0s3server branch branch to deploy'
-         default: 'staging'
-         required: true
-       zbox_branch:
-        description: '0box branch to deploy'
-        default: 'staging'
+       repo_snapshots_branch:
```
Comment (dabasov) [inline .github/workflows/system_tests.yml]: and this one

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | sprint-july-2  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | fix/finishAllocation  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | fix/finishAllocation  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 4]
```diff
@@ -1308,6 +1304,9 @@ func (sc *StorageSmartContract) finalizedPassRates(alloc *StorageAllocation) ([]
 	alloc.Stats.SuccessChallenges = succesful
 	alloc.Stats.FailedChallenges = failed
 	alloc.Stats.OpenChallenges = 0
+
+	emitUpdateAllocationAndBlobberStats(alloc, balances)
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:1308]: can we emit it higher the stack where `finalizedPassRates` is called?

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 5]
```diff
@@ -1371,11 +1380,13 @@ func (sc *StorageSmartContract) canceledPassRates(
 			continue
 		}
 		// success rate for the blobber allocation
-		// fmt.Println("pass rate i", i, "successful", d.Stats.SuccessChallenges, "failed", d.Stats.FailedChallenges)
 		passRates = append(passRates, float64(ba.Stats.SuccessChallenges)/float64(ba.Stats.TotalChallenges))
 	}
 
 	alloc.Stats.OpenChallenges = 0
+
+	emitUpdateAllocationAndBlobberStats(alloc, balances)
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:1388]: also maybe we can call it higher the stack

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 6]
```diff
@@ -1605,16 +1609,20 @@ func (sc *StorageSmartContract) finishAllocation(
...
+	maxChallengeCompletionDTU := float64(conf.MaxChallengeCompletionTime) / float64(conf.TimeUnit)
+	adjustableChallengePoolTokens := cpBalance * maxChallengeCompletionDTU
+
 	var passPayments currency.Coin
 	for i, d := range alloc.BlobberAllocs {
-		if alloc.UsedSize > 0 && cp.Balance > 0 && passRates[i] > 0 && d.Stats != nil {
-			ratio := float64(d.Stats.UsedSize) / float64(alloc.UsedSize)
-			cpBalance, err := cp.Balance.Float64()
+		if alloc.Stats.UsedSize > 0 && cp.Balance > 0 && passRates[i] > 0 && d.Stats != nil {
+			ratio := float64(d.Stats.UsedSize) / (float64(alloc.Stats.UsedSize) * float64(alloc.DataShards+alloc.ParityShards) / float64(alloc.DataShards))
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: can you please break it into 2 different ratios and then multiply them?

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/allocation.go (hunk 4 above)
Comment (Jayashsatolia403) [reply code/go/0chain.net/smartcontract/storagesc/allocation.go:1308]: Not sure why do we want to do this. Can you please explain?

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**